### ISSUE TITLE
Worker nodes are stuck up in inspecting status.

### ISSUE THREAD

**Issue by maveric-tellrv** [openshift-metal3/dev-scripts#761]
**Describe the bug**
Worker nodes are stuck up in inspecting status.
```
oc -n openshift-machine-api get bmh
NAME                 STATUS   PROVISIONING STATUS      CONSUMER                BMC                         HARDWARE PROFILE   ONLINE   ERROR
openshift-master-0   OK       externally provisioned   ostest-master-0         ipmi://192.168.111.1:6230                      true     
openshift-master-1   OK       externally provisioned   ostest-master-1         ipmi://192.168.111.1:6231                      true     
openshift-master-2   OK       externally provisioned   ostest-master-2         ipmi://192.168.111.1:6232                      true     
openshift-worker-0   OK       inspecting               ostest-worker-0-4sqbm   ipmi://192.168.111.1:6233                      true     
openshift-worker-1   OK       inspecting               ostest-worker-0-n5ds2   ipmi://192.168.111.1:6234                      true     
openshift-worker-2   OK       inspecting               ostest-worker-0-6rmdj   ipmi://192.168.111.1:6235                      true     
``

**To Reproduce**
Install using latest dev-script via ' make' command

**Expected/observed behavior**
I have modified the config_user.sh file by using below env variables to launch 3 worker node
export NUM_MASTERS="3"
export NUM_WORKERS="3"
export VM_EXTRADISKS="true"

but after the installation is complete on master nodes comes up .
```[rhhi@titan98 dev-scripts]$ oc get nodes
NAME       STATUS   ROLES           AGE    VERSION
master-0   Ready    master,worker   173m   v1.14.0+a80442411
master-1   Ready    master,worker   173m   v1.14.0+a80442411
master-2   Ready    master,worker   173m   v1.14.0+a80442411
```
**Additional context**
```
 oc -n openshift-machine-api get bmh
NAME                 STATUS   PROVISIONING STATUS      CONSUMER                BMC                         HARDWARE PROFILE   ONLINE   ERROR
openshift-master-0   OK       externally provisioned   ostest-master-0         ipmi://192.168.111.1:6230                      true     
openshift-master-1   OK       externally provisioned   ostest-master-1         ipmi://192.168.111.1:6231                      true     
openshift-master-2   OK       externally provisioned   ostest-master-2         ipmi://192.168.111.1:6232                      true     
openshift-worker-0   OK       inspecting               ostest-worker-0-4sqbm   ipmi://192.168.111.1:6233                      true     
openshift-worker-1   OK       inspecting               ostest-worker-0-n5ds2   ipmi://192.168.111.1:6234                      true     
openshift-worker-2   OK       inspecting               ostest-worker-0-6rmdj   ipmi://192.168.111.1:6235                      true     

[rhhi@titan98 ~]$ oc get pods  -n openshift-machine-api
NAME                                          READY   STATUS    RESTARTS   AGE
cluster-autoscaler-operator-5ff8484fc-zl26n   1/1     Running   0          139m
machine-api-controllers-76b7bc94b5-s44vt      3/3     Running   0          159m
machine-api-operator-66787f5866-nw82x         1/1     Running   1          161m
metal3-5699969bcd-xz4k2                       8/8     Running   0          158m

[rhhi@titan98 ~]$ oc -n openshift-machine-api logs metal3-5699969bcd-xz4k2
Error from server (BadRequest): a container name must be specified for pod metal3-5699969bcd-xz4k2, choose one of: [metal3-baremetal-operator metal3-dnsmasq metal3-mariadb metal3-httpd metal3-ironic-conductor metal3-ironic-api metal3-ironic-inspector metal3-static-ip-manager] or one of the init containers: [metal3-ipa-downloader metal3-rhcos-downloader metal3-static-ip-set]


[rhhi@titan98 ~]$ export OS_TOKEN=fake-token
[rhhi@titan98 ~]$ export OS_URL=http://172.22.0.2:6385/
[rhhi@titan98 ~]$ openstack baremetal node list
Error contacting Ironic server: Unable to establish connection to http://172.22.0.2:6385/v1/nodes: HTTPConnectionPool(host='172.22.0.2', port=6385): Max retries exceeded with url: /v1/nodes (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f2926997590>: Failed to establish a new connection: [Errno 113] No route to host',)). Attempt 6 of 6
Unable to establish connection to http://172.22.0.2:6385/v1/nodes: HTTPConnectionPool(host='172.22.0.2', port=6385): Max retries exceeded with url: /v1/nodes (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f2926997590>: Failed to establish a new connection: [Errno 113] No route to host',))



```

**Comment by hardys** [comment]
> export OS_URL=http://172.22.0.2:6385/

This should be export OS_URL=http://172.22.0.3:6385/ after the bootstrap VM is deleted, see:

https://github.com/openshift-metal3/dev-scripts#interacting-with-ironic-directly

It would be helpful to check the primary console of the nodes stuck inspecting - did the ramdisk boot OK?  Any clues in the logs?

**Comment by maveric-tellrv** [comment]
@hardys , I am not able to login any of the worker node . Any pointer which logs should i check more ? 
```
[rhhi@titan98 ~]$ export OS_TOKEN=fake-token
[rhhi@titan98 ~]$ export OS_URL=http://172.22.0.3:6385
[rhhi@titan98 ~]$ openstack baremetal node list
+--------------------------------------+--------------------+--------------------------------------+-------------+--------------------+-------------+
| UUID                                 | Name               | Instance UUID                        | Power State | Provisioning State | Maintenance |
+--------------------------------------+--------------------+--------------------------------------+-------------+--------------------+-------------+
| 14b33683-13c4-41fa-b923-3905151eca28 | openshift-master-0 | 14b33683-13c4-41fa-b923-3905151eca28 | power on    | active             | False       |
| 8f6ed9d4-c410-4511-bba4-a91b116d9d3d | openshift-master-1 | 8f6ed9d4-c410-4511-bba4-a91b116d9d3d | power on    | active             | False       |
| 5e3d458a-418d-429c-82f6-7ea5232c3510 | openshift-master-2 | 5e3d458a-418d-429c-82f6-7ea5232c3510 | power on    | active             | False       |
| fe928f88-c02d-46bf-accf-52e4f1c9c408 | openshift-worker-0 | None                                 | power on    | manageable         | False       |
| 0bf3e708-7498-4cc2-8052-0910569e31bc | openshift-worker-1 | None                                 | power on    | manageable         | False       |
| d00b5698-3563-44de-a02a-6bc108a13ac4 | openshift-worker-2 | None                                 | power on    | manageable         | False       |
+--------------------------------------+--------------------+--------------------------------------+-------------+--------------------+-------------+
[rhhi@titan98 ~]$ sudo virsh list --all 
 Id    Name                           State
----------------------------------------------------
 127   openshift_master_1             running
 128   openshift_master_0             running
 129   openshift_master_2             running
 131   openshift_worker_1             running
 132   openshift_worker_2             running
 133   openshift_worker_0             running

[rhhi@titan98 ~]$ sudo virsh console openshift_worker_0 
Connected to domain openshift_worker_0
Escape character is ^]

```
Screenshot of one of the worker-node from virt-manager
![Screenshot from 2019-08-22 15-27-33](https://user-images.githubusercontent.com/20462934/63507436-598daf00-c4f5-11e9-9354-9dafba9c4852.png)

**Comment by maveric-tellrv** [comment]
Adding deployment logs 
[11_register_hosts-2019-08-22-093320.log](https://github.com/openshift-metal3/dev-scripts/files/3530128/11_register_hosts-2019-08-22-093320.log)
[add-machine-ips-2019-08-22-093324.log](https://github.com/openshift-metal3/dev-scripts/files/3530130/add-machine-ips-2019-08-22-093324.log)
[12_csr_hack-2019-08-22-093324.log](https://github.com/openshift-metal3/dev-scripts/files/3530133/12_csr_hack-2019-08-22-093324.log)
[06_create_cluster-2019-08-22-084035.log](https://github.com/openshift-metal3/dev-scripts/files/3530135/06_create_cluster-2019-08-22-084035.log)
[04_setup_ironic-2019-08-22-083945.log](https://github.com/openshift-metal3/dev-scripts/files/3530136/04_setup_ironic-2019-08-22-083945.log)
[03_ocp_repo_sync-2019-08-22-083944.log](https://github.com/openshift-metal3/dev-scripts/files/3530138/03_ocp_repo_sync-2019-08-22-083944.log)
[02_configure_host-2019-08-22-083840.log](https://github.com/openshift-metal3/dev-scripts/files/3530142/02_configure_host-2019-08-22-083840.log)
[01_install_requirements-2019-08-22-083744.log](https://github.com/openshift-metal3/dev-scripts/files/3530145/01_install_requirements-2019-08-22-083744.log)

**Comment by russellb** [comment]
I've reproduced this.  In my case, I see in the console log that inspection finished, but then it says: `No ipa-api-url configured, Heartbeat and lookup skipped for inspector.`

**Comment by sadasu** [comment]
I am also hitting this but see this error message in the metal3-ironic-inspector container logs:

2019-08-22 15:41:52.879 1 ERROR ironic_inspector.plugins.extra_hardware [req-5693a099-1e8f-42d7-9b9d-34eb295cd869 - - - - -] [node: 3e782cbe-48a3-443c-8584-63848333523f state processing] Failed to save extra hardware information in Swift: Could not create an adapter to connect to the object storage service: Auth plugin requires parameters which were not given: auth_url: ironic_inspector.utils.Error: Could not create an adapter to connect to the object storage service: Auth plugin requires parameters which were not given: auth_url^[[00m

**Comment by sadasu** [comment]
Follow up: After the above error, the logs contain:
2019-08-22 15:41:52.885 1 DEBUG ironic_inspector.plugins.introspection_data [req-5693a099-1e8f-42d7-9b9d-34eb295cd869 - - - - -] Introspection data storage is disabled, the data will not be saved for node 3e782cbe-48a3-443c-8584-63848333523f save /usr/lib/python3.6/site-packages/ironic_inspector/plugins/introspection_data.py:76

After this point, the logs say this:
2019-08-22 15:41:52.932 1 INFO ironic_inspector.process [-] [node: 3e782cbe-48a3-443c-8584-63848333523f state finished MAC 00:42:76:eb:a0:cd] Introspection finished successfully^[[00m

But, then this is what happens after:
2019-08-22 15:42:05.080 1 ERROR ironic_inspector.utils [req-e2ae9b3a-5bdd-4ac4-abfa-3183620d4e2b - - - - -] Introspection data storage is disabled^[[00m
2019-08-22 15:42:05.081 1 DEBUG ironic_inspector.main [req-e2ae9b3a-5bdd-4ac4-abfa-3183620d4e2b - - - - -] Returning error to client: Inspector is not configured to store data. Set the [processing]store_data configuration option to change this. error_response /usr/lib/python3.6/site-packages/ironic_inspector/main.py:71

**Comment by russellb** [comment]
> 2019-08-22 15:42:05.081 1 DEBUG ironic_inspector.main [req-e2ae9b3a-5bdd-4ac4-abfa-3183620d4e2b - - - - -] Returning error to client: Inspector is not configured to store data. Set the [processing]store_data configuration option to change this. error_response /usr/lib/python3.6/site-packages/ironic_inspector/main.py:71

This is due to turning off inspecting data:

https://github.com/openshift/ironic-inspector-image/pull/13

It may be that we have to get that fixed before we can proceed with successful worker deployments again.

**Comment by rdoxenham** [comment]
> Follow up: After the above error, the logs contain:
> 2019-08-22 15:41:52.885 1 DEBUG ironic_inspector.plugins.introspection_data [req-5693a099-1e8f-42d7-9b9d-34eb295cd869 - - - - -] Introspection data storage is disabled, the data will not be saved for node 3e782cbe-48a3-443c-8584-63848333523f save /usr/lib/python3.6/site-packages/ironic_inspector/plugins/introspection_data.py:76
> 
> After this point, the logs say this:
> 2019-08-22 15:41:52.932 1 INFO ironic_inspector.process [-] [node: 3e782cbe-48a3-443c-8584-63848333523f state finished MAC 00:42:76:eb:a0:cd] Introspection finished successfully^[[00m
> 
> But, then this is what happens after:
> 2019-08-22 15:42:05.080 1 ERROR ironic_inspector.utils [req-e2ae9b3a-5bdd-4ac4-abfa-3183620d4e2b - - - - -] Introspection data storage is disabled^[[00m
> 2019-08-22 15:42:05.081 1 DEBUG ironic_inspector.main [req-e2ae9b3a-5bdd-4ac4-abfa-3183620d4e2b - - - - -] Returning error to client: Inspector is not configured to store data. Set the [processing]store_data configuration option to change this. error_response /usr/lib/python3.6/site-packages/ironic_inspector/main.py:71

Yeah, so this is to be expected currently - the inspection service attempts to upload the data (which is the default) but inspection data storage is disabled on the default container image for inspector. See here: https://github.com/openshift/ironic-inspector-image/blob/master/inspector.conf#L22

We're hoping to bring that back to enabled ASAP (https://github.com/openshift/ironic-inspector-image/pull/13). The problem here is that the BMO **requires** some inspection data to be returned by Ironic before it can shift to a 'ready' state. You can hack this with this PR: https://github.com/openshift-metal3/dev-scripts/pull/751.

Alternatively you can use a custom release that has store_data enabled (quay.io/roxenham/ironic-inspector-image as an example as that's what I'm using).

**Comment by hardys** [comment]
> > 2019-08-22 15:42:05.081 1 DEBUG ironic_inspector.main [req-e2ae9b3a-5bdd-4ac4-abfa-3183620d4e2b - - - - -] Returning error to client: Inspector is not configured to store data. Set the [processing]store_data configuration option to change this. error_response /usr/lib/python3.6/site-packages/ironic_inspector/main.py:71
> 
> This is due to turning off inspecting data:
> 
> [openshift/ironic-inspector-image#13](https://github.com/openshift/ironic-inspector-image/pull/13)
> 
> It may be that we have to get that fixed before we can proceed with successful worker deployments again.

My local testing indicates this now works (only tested via podman so far though), so we may want to consider reverting that, then re-testing e2e with a CI build to prove all is OK, with the option to revert before the next nightly if there are any issues.

**Comment by hardys** [comment]
This was fixed for openshift/ironic-inspector-image#17  for the images that dev-scripts consumes by default (e.g not the CI builds, the nightly OCP ones) so closing this issue, the openshift/release links in the comment above can be used to track resolution of the issues with the CI builds.
